src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -737,15 +736,37 @@ AttributeEquivalenceClassForEquivalenceClass(EquivalenceClass *plannerEqClass,
 
 		if (IsA(strippedEquivalenceExpr, Param))
 		{
-			List *parentParamList = relationRestriction->parentPlannerParamList;
+			PlannerInfo *outerNodeRoot = relationRestriction->plannerInfo->parent_root;
+			ListCell *outerPlanParamListCell = NULL;
 			Param *equivalenceParam = (Param *) strippedEquivalenceExpr;
 
-			expressionVar = GetVarFromAssignedParam(parentParamList,
-													equivalenceParam);
+			/*
+			 * A Param can reference a Var in any of the outer plan nodes. Search
+			 * them until we find the corresponding Var.
+			 */
+			foreach(outerPlanParamListCell, relationRestriction->outerPlannerParamLists)
+			{
+				List *outerPlanParamList = lfirst(outerPlanParamListCell);
+
+				/*
+				 * outerPlannerParamLists has the same number of elements as the
+				 * depth of relationRestriction->plannerInfo.
+				 */
+				Assert(outerNodeRoot != NULL);
+
+				expressionVar = SearchPlannerParamList(outerPlanParamList,
+													   equivalenceParam);
+				if (expressionVar != NULL)
+				{
+					break;
+				}
+
+				outerNodeRoot = outerNodeRoot->parent_root;
+			}
+
 			if (expressionVar)
 			{
-				AddToAttributeEquivalenceClass(&attributeEquivalance,
-											   relationRestriction->parentPlannerInfo,
+				AddToAttributeEquivalenceClass(&attributeEquivalance, outerNodeRoot,
```

Comment:
Hmm, this is tricky to follow and expand in the future because we're trying to match the `expressionVar ` and its corresponding `root` via the loop. I think it is OK for this change, but, for any future users of these fields, it'd be pretty hard to follow the dependency and prone to bugs.

What I had in my mind was to have a new struct and pass the params and its corresponding `root` together and use them together. 

```C
typedef struct RootParams
{
   PlannerInfo *root;
  List *plan_params
};

```

What do you think @pykello ?


(the struct name and field names are placeholders, feel free to improve.)
(Also note that we don't need to copy `root` field, only the params as you did.)

Reply:
makes sense, done.